src/main/scala/shine/DPIA/Types/DataType.scala

```diff
@@ -12,6 +12,67 @@ sealed trait BasicType extends DataType
 
 sealed trait ScalarType extends BasicType
 
+sealed trait MatrixLayout
+
+object MatrixLayout {
+  object Row_Major extends MatrixLayout { override def toString = "Row_Major" }
+  object Col_Major extends MatrixLayout { override def toString = "Col_Major" }
+}
+
+final case class MatrixLayoutIdentifier(name: String) extends MatrixLayout with Kind.Identifier {
+  var layout: Option[MatrixLayout] = None
```

Comment:
should the identifier not be substituted after inference, instead of mutating an internal variable?

Reply:
The `wmma::fragment<...>`-Type needs to know the layout of the matrix in the memory (the matrix can be stored row by row or column by column).
We do not want do specify the layout manually. So it should be inferred.
In the current version the layout is inferred when we generate code for the imperative `WmmaLoad`-primitive, which loads a matrix (represented as `ArrayType(numberOfRows, ArrayType(numberOfColumns, datatype))`) into a fragment. When we want to generate code for this primitive, we also generate code for different index accesses for passed matrix, so we can decide whether the matrix is stored row by row or column by column in the memory.
But this happens not until code generation for the primitive. So the functional and imperativ expressions still contains this not inferred Identifier.
Before the identifier is inferred, it is possible that we insert statements in the AST, which are using this identifier.
That results in this Type definitions:
https://github.com/rise-lang/shine/blob/e1b826085f626c05541f2cbb20452532c87552df/src/main/scala/shine/cuda/ast/Types.scala#L22-L31
It could be a good idea to infer the layout (and the leading dimension, which describes the number of elements between to rows of the matrix in the memory) when we have the rise expression, cause we usually do not want such identifier in our DPIA-Code
It should be possible to do this, but i do not know how to do this.

Reply:
Issue https://github.com/rise-lang/shine/issues/111